### ISSUE TITLE
[FEA] Port remaining NVText APIs to libcudf++

### ISSUE THREAD

**Issue by davidwendt** [rapidsai/cudf#5243]
Some of the NVText APIs from the custrings/nvstrings package were ported to libcudf++ as requested per issue #3965

Here are a list of the APIs from the original NVText class
https://docs.rapids.ai/api/libnvstrings/stable/classNVText.html

Here are a list of the remaining APIs. The links are to the python documentation which has examples. The C++ documentation can be seen in the class link above.

- [ ] [unique_tokens](https://docs.rapids.ai/api/nvstrings/stable/api.html#nvtext.unique_tokens)
- [x] [character_tokenize](https://docs.rapids.ai/api/nvstrings/stable/api.html#nvtext.character_tokenize)
- [ ] [contains_strings](https://docs.rapids.ai/api/nvstrings/stable/api.html#nvtext.contains_strings)
- [ ] [strings_counts](https://docs.rapids.ai/api/nvstrings/stable/api.html#nvtext.strings_counts)
- [x] [tokens_counts](https://docs.rapids.ai/api/nvstrings/stable/api.html#nvtext.tokens_counts)
- [x] [replace_tokens](https://docs.rapids.ai/api/nvstrings/stable/api.html#nvtext.replace_tokens)
- [x] [edit_distance](https://docs.rapids.ai/api/nvstrings/stable/api.html#nvtext.edit_distance)
- [x] [edit_distance_matrix](https://docs.rapids.ai/api/nvstrings/stable/api.html#nvtext.edit_distance_matrix)
- [x] is_letter: [consonant](https://docs.rapids.ai/api/nvstrings/stable/api.html#nvtext.is_consonant), [vowel](https://docs.rapids.ai/api/nvstrings/stable/api.html#nvtext.is_vowel)
- [x] [porter_stemmer_measure](https://docs.rapids.ai/api/nvstrings/stable/api.html#nvtext.porter_stemmer_measure)
- [ ] [scatter_count](https://docs.rapids.ai/api/nvstrings/stable/api.html#nvtext.scatter_count)

I'm not sure if all the NVText APIs are still being used.
Some of these were created to just try out different algorithms and then were abandoned when other techniques were used instead.

@randerzander @beckernick @VibhuJawa

**Comment by VibhuJawa** [comment]
Of these i can confirm that below will be used by [count-vectorizer/hashing-vectorizer](https://github.com/rapidsai/cuml/pull/2267): 

- [x] unique_tokens
- [x] character_tokenize
- [x] strings_counts
- [x] tokens_counts
- [x] replace_tokens

Below is useful to have for string workflows but i am not sure if its being actively used anywhere : 
- [ ] contains_strings

For `scatter_count` we use [repeat](https://github.com/rapidsai/cudf/blob/839bed5b1a54adce49f3aa6ffaf11d8d83efc5c2/python/cudf/cudf/core/frame.py#L978-L988
) instead  for which we do the same thing. 
- [ ]   scatter_count

Below might be useful for `bio-informatics/spell correction` workflows , which we might start looking at soon: 

- [ ]   edit_distance
- [ ]   edit_distance_matrix


Below are only useful for porter stemmer once we finally create it (if we still want to ) 
- [ ]   is_letter: consonant, vowel
- [ ]   porter_stemmer_measure

**Comment by VibhuJawa** [comment]
Also CC: @Garfounkel , who is working at `tf-idf` transformer so he might have thoughts on using some of these functions.

**Comment by Garfounkel** [comment]
For [CountVectorizer](https://github.com/rapidsai/cuml/pull/2267) and `HashingVectorizer` we would also need a way to generate ngrams on a character level, there is a feature request [here](https://github.com/rapidsai/cudf/issues/5144). This would enable a better `TfidfVectorizer` as well.

**Comment by kkraus14** [comment]
@davidwendt if possible could you prioritize `replace_tokens`, `ngram_tokenize`, and `character_tokenize` to unblock https://github.com/rapidsai/cuml/pull/2267/files as we'll be removing nvtext from 0.15 shortly.

**Comment by VibhuJawa** [comment]
> @davidwendt if possible could you prioritize `replace_tokens`, `ngram_tokenize`, and `character_tokenize` to unblock https://github.com/rapidsai/cuml/pull/2267/files as we'll be removing nvtext from 0.15 shortly.

Small clarification: `ngram_tokenize` is all ready there `replace_tokens` and `character_tokenize` will need switching.  


Below has changed now as the `pr` has changed implementations a bit from the initial work. 
> Of these i can confirm that below will be used by [count-vectorizer/hashing-vectorizer](https://github.com/rapidsai/cuml/pull/2267):
> 
> * [x]  unique_tokens
> * [ ]  character_tokenize
> * [x]  strings_counts
> * [x]  tokens_counts
> * [x]  replace_tokens
 
Now it only uses: 
 
 * [ ]  unique_tokens
 * [x]  character_tokenize
 * [ ]  strings_counts
 * [ ]  tokens_counts
* [x]  replace_tokens
 
Sorry for the confusion around this, @Garfounkel  please feel free to correct this.

**Comment by kkraus14** [comment]
I don't see anywhere in the code it's using `unique_tokens`, `strings_counts` or `tokens_counts` currently. Could you point this out?

**Comment by VibhuJawa** [comment]
> I don't see anywhere in the code it's using `unique_tokens`, `strings_counts` or `tokens_counts` currently. Could you point this out?

I meant to say it only uses , `character_tokenize ` and `replace_tokens ` , have  marked them with `ticks` . 

 Sorry again. :-(

**Comment by Garfounkel** [comment]
@davidwendt Thank you very much for your work on those, it is really helpful to me and will unblock NLP estimators for cuML which is very exciting.

I am not sure if this is out of scope for this issue but I noticed that [replace_multi](https://docs.rapids.ai/api/nvstrings/stable/api.html#nvstrings.nvstrings.replace_multi) from nvstrings is also missing from `Series.str`. This is an other function needed to unblock https://github.com/rapidsai/cuml/pull/2267/files.

**Comment by davidwendt** [comment]
@Garfounkel I believe `replace_multi` just got overloaded into `Series.str.replace()`
https://docs.rapids.ai/api/cudf/stable/api.html#cudf.core.column.string.StringMethods.replace
The parameters here handle a single string or column(s) of strings.

**Comment by Garfounkel** [comment]
I see, you're right, thank you!

**Comment by davidwendt** [comment]
PR #5811 added the `edit_distance` API to [libcudf](https://docs.rapids.ai/api/libcudf/nightly/group__nvtext__edit__distance.html) libcudf and [cudf](https://docs.rapids.ai/api/cudf/nightly/api.html#cudf.core.column.string.StringMethods.edit_distance)

@VibhuJawa I think the only thing left here is `unique_tokens`. Is that something we should port? Otherwise we can probably close this and open a new feature request if we find we need it later on.
cc @randerzander

**Comment by VibhuJawa** [comment]
> PR #5811 added the `edit_distance` API to [libcudf](https://docs.rapids.ai/api/libcudf/nightly/group__nvtext__edit__distance.html) libcudf and [cudf](https://docs.rapids.ai/api/cudf/nightly/api.html#cudf.core.column.string.StringMethods.edit_distance)
> 
> @VibhuJawa I think the only thing left here is `unique_tokens`. Is that something we should port? Otherwise we can probably close this and open a new feature request if we find we need it later on.

I don't think we need it currently, Good to close from my side.

**Comment by kkraus14** [comment]
Closing!
